### ISSUE TITLE
Bar menus

### ISSUE THREAD

**Issue by bustawin** [r0x0r/pywebview#500]
### Description  
Although this question is not strictly about pywebview if I can get some help I could write a guide with the code or even work with a PR in the future if there is interest.

I have seen that a menu bar is quite required for my app (ex. File > Open, File > Save, Edit > ...). Although I plan to release it in the three platforms, I am starting with Mac, which let me try to hijack the Cocoa's BrowserView._add_view_menu to add simple menu items that would ultimately just call a custom python function.

I have never touched PyObjC. Just mimicking the existing code I can create new menus
`myMenu.addItemWithTitle_action_keyEquivalent_("full", "foobar:", "o")` but I don't know how to do it to execute my python function (I understand I need to add a foobar_() function somewhere...). PyObjC guides / example code have not helped me, as there are differences between the ones I have seen and your implementation. Could I get some help?

I understand in Linux / Windows this task will less cryptic?

Thank you!

Btw, amazing tool you have here :-)
  
### Practicalities

- YES I am willing to work on this issue myself.
- NO I am prepared to support this issue financially.

**Comment by r0x0r** [comment]
Objective C is confusing af and can be intimidating. PyObjC adds its own quirks, but the good news is that it mostly works, once you get the syntax. As for the menu, somebody else has implemented it, so myself I have only a vague idea how it should be done. `addItemWithTitle_action_keyEquivalent_` seems to be the right method and action is attached via a selector.  Here is an example of how to use a selector in Python. https://lethain.com/how-to-use-selectors-in-pyobjc/


The menubar has been requested several times and in my opinion it can be incorporated into `pywebview`. If you would like to extend your work into a generic solution, that would be most welcomed. The scaffold for the API would be something like this 
```
menu_items = {
  'File': (
    ('First item', action),
    (), // separator
    ('Second item', action2),
  )
  'Edit': ( ....)
}
menu = webview.create_menu(menu_items)
window = webview.create_window('Window with menu', menu=menu)
```

**Comment by bustawin** [comment]
Thank you! I will take a look.

I will see if I can make something generic work for my project and come back.

**Comment by bustawin** [comment]
I have been able to create a working menu :-)

Now the problem is that I don't know how to relate the menu to the active window. For now, the menu is global. So if I click "save" and I have several windows I don't know which window to save.
Ideally the `my_action()` should receive a `window` argument to know which window executed it.
 Any idea how to get it?

```python
from __future__ import annotations

import typing as t

import AppKit
import objc
from webview.platforms.cocoa import BrowserView


def add_view(self: BrowserView):
    super_view_menu(self)

    m = Menu("File", parent=self.app.mainMenu())
    m.action("First item", my_action, command="F")
    (
        m.sub_menu("Second item")
        .action("First sub item", my_action)
        .separator()
        .action("Second sub item", my_action, command="R")
    )


super_view_menu = BrowserView._add_view_menu
BrowserView._add_view_menu = add_view


def my_action():
    print("executed!")


class Menu:
    def __init__(self, title: str, parent: t.Union[Menu, "NSMenu"]):
        self.m = AppKit.NSMenu.alloc().init()
        self.item = AppKit.NSMenuItem.alloc().init()
        self.item.setSubmenu_(self.m)
        if not isinstance(parent, self.__class__):
            self.m.setTitle_(title)
            parent.addItem_(self.item)
        else:
            self.item.setTitle_(title)
            parent.m.addItem_(self.item)

    def action(self, title: str, action: callable, command: t.Optional[str] = None):
        Action(self, title, action, command)
        return self

    def separator(self):
        self.m.addItem_(AppKit.NSMenuItem.separatorItem())
        return self

    def sub_menu(self, title: str):
        return self.__class__(title, parent=self)


class Action:
    def __init__(self, parent: Menu, title: str, action: callable, command=None):
        self.action = action
        s = objc.selector(self._call_action, signature=b"v@:")
        if command:
            item = parent.m.addItemWithTitle_action_keyEquivalent_(title, s, command)
        else:
            item = AppKit.NSMenuItem.alloc().init()
            item.setAction_(s)
            item.setTitle_(title)
            parent.m.addItem_(item)
        item.setTarget_(self)

    def _call_action(self):
        self.action()



```

**Comment by AlexanderWillner** [comment]
> As for the menu, somebody else has implemented it,

This was @cuibonobo btw
> I have been able to create a working menu :-)

Great. Started to look into this to add a "Preferences" item, however, not succeeded yet.

**Comment by Ksengine** [comment]
you can use html to menu and iframe to webpage

**Comment by shepazon** [comment]
> you can use html to menu and iframe to webpage

The down side to that is that doesn't get you a proper menu bar. You can make something that looks like one, but since the Mac puts all menu bars at the top of the screen, instead of inside windows, the result is very obviously non-native UI.

**Comment by r0x0r** [comment]
@sardination's PR is merged into the 4.0 branch. Testing and QA would be much appreciated.
